File: 장세훈/LeetCode/medium/number of islands.java
class NumberOfIslands {
    /*
        200. Number of Islands
        https://leetcode.com/problems/number-of-islands/description/

        섬의 개수 세는 문제
        stack을 활용하 bfs로 문제를 풀이함.
          2차원 배열을 순회하면서 1이 나오면
            answer++
            1의 위치를 기준으로 상하좌우 위치를 스택에 넣고 뺴며 섬을 지워나감

     */

    public int numIslands(char[][] grid) {
        int n = grid.length;
        int m = grid[0].length;

        int answer = 0;
        int[][] dir = new int[][]{{1, 0}, {0, -1}, {-1, 0}, {0, 1}};


        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == '1') {
                    answer++;
                    Deque<int[]> stack = new ArrayDeque<>();
                    stack.push(new int[]{i, j});

                    while (!stack.isEmpty()) {
                        int[] pop = stack.pop();
                        grid[pop[0]][pop[1]] = '0';

                        for (int k = 0; k < 4; k++) {
                            int nx = dir[k][0] + pop[0];
                            int ny = dir[k][1] + pop[1];

                            if (nx > -1 && nx < n && ny > -1 && ny < m && grid[nx][ny] == '1') {
                                grid[nx][ny] = '0';
                                stack.add(new int[]{nx, ny});
                            }
                        }
                    }
                }
            }
        }

        return answer;
    }
}

File: 정수빈/LeetCode/medium/number-of-islands.py
# https://leetcode.com/problems/number-of-islands/
# 설명: dfs, bfs문제 
# 해결못함
# 프로그래머스에서 푼 문제인데 동일한 문제인데도 못풀었네요 ㅠㅠ 정말 bfs, dfs는 어려운거 같습니다. ㅠㅠㅠ

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0
        
        m, n = len(grid), len(grid[0])
        from collections import deque


        visited = [[False for _ in range(n)] for _ in range(m)]
        offset = [(1, 0), (0, 1), (-1,0), (0, -1)]


        def dfs(y_, x_):
            print("y_, x_",y_, x_)
            queue = deque([(y_,x_)])
            visited[y_][x_] = True

            while queue:
              y, x = queue.popleft()

              for dy, dx in offset:
                nx = x + dx
                ny = y + dy
                if 0 <= nx < n and 0 <= ny < m:
                    if not visited[ny][nx] and grid[ny][nx] == "1":
                        queue.append((ny, nx))
                        visited[ny][nx] = True

        ans = 0
        for y in range(m):
            for x in range(n):
                if grid[y][x] == "1" and not visited[y][x]:
                    print(y, x)
                    dfs(y, x)
                    ans +=1
        return ans
                    

